query/src/main/java/org/apache/kylin/query/util/PushDownExecutor.java

```diff
@@ -0,0 +1,132 @@
+            boolean isPrepare) throws Exception {
+        List<String> ids = kylinConfig.getPushDownRunnerIds();
+        if (ids.isEmpty()) {
+            IPushDownRunner runner = (IPushDownRunner) ClassUtil.newInstance(
+                    kylinConfig.getPushDownRunnerClassName()
+            );
+            runner.init(kylinConfig);
+            return queryBySingleRunner(runner, project, sql, defaultSchema, sqlException,
+                    isSelect, isPrepare);
+        } else {
+            return queryByMultiJdbcRunners(ids, project, sql, defaultSchema, sqlException,
+                    isSelect, isPrepare);
+        }
+    }
+
+    private Pair<List<List<String>>, List<SelectedColumnMeta>> queryBySingleRunner(IPushDownRunner runner,
+            String project, String sql, String defaultSchema, SQLException sqlException,
+            boolean isSelect, boolean isPrepare) throws Exception {
+
+        logger.debug("Query Pushdown runner {}", runner);
+
+        // default schema in calcite does not apply to other engines.
+        // since this is a universql requirement, it's not implemented as a converter
+        if (defaultSchema != null && !defaultSchema.equals("DEFAULT")) {
+            String completed = sql;
+            try {
+                completed = PushDownUtil.schemaCompletion(sql, defaultSchema);
+            } catch (SqlParseException e) {
+                // fail to parse the pushdown sql, ignore
+                logger.debug("fail to do schema completion on the pushdown sql, ignore it.",
+                        e.getMessage());
+            }
+            if (!sql.equals(completed)) {
+                logger.info("the query is converted to {} after schema completion", completed);
+                sql = completed;
+            }
+        }
+
+        sql = runner.convertSql(kylinConfig, sql, project, defaultSchema, isPrepare);
+
+        List<List<String>> returnRows = Lists.newArrayList();
+        List<SelectedColumnMeta> returnColumnMeta = Lists.newArrayList();
+
+        if (isSelect) {
+            runner.executeQuery(sql, returnRows, returnColumnMeta);
+        }
+        if (!isSelect && !isPrepare && kylinConfig.isPushDownUpdateEnabled()) {
+            runner.executeUpdate(sql);
+        }
+        return Pair.newPair(returnRows, returnColumnMeta);
+    }
+
+    private Pair<List<List<String>>, List<SelectedColumnMeta>> queryByMultiJdbcRunners(
+            List<String> ids, String project, String sql, String defaultSchema,
+            SQLException sqlException, boolean isSelect, boolean isPrepare) throws Exception {
+        Pair<List<List<String>>, List<SelectedColumnMeta>> ret = null;
+
+        for (int i = 0; i < ids.size(); i++) {
+            String id = ids.get(i);
+            PushDownRunnerJdbcImpl runner = new PushDownRunnerJdbcImpl();
+            runner.initById(kylinConfig, id);
+
+            try {
+                ret = queryBySingleRunner(runner, project, sql, defaultSchema, sqlException,
```

Comment:
Why not return the result from the first engine immediately if more than one engines are available?

Reply:
@nichunen , thanks to review, it has been fixed